descriptors: Allow `H` as a hardened indicator

achow101 opened the pull request:
BIP 380 states that `H` is a valid hardened indicator.

achow101 committed:
descriptors: Allow H as a hardened indicator

BIP 380 states that H is a valid hardened indicator.

jonatack commented:
Concept ACK, see https://github.com/bitcoin/bips/pull/1803

pablomartin4btc reviewed (commented):
ACK b6252655fe502ed1bbb97ab9597d9bbcf8c4484c

Dorex45 reviewed (commented):
Great update! Supporting both 'H' and 'h' for hardened keys makes things easier for everyone and helps avoid confusion. The extra tests are a nice touch too. Thanks for making this even more user-friendly!

janb84 reviewed (commented):
ACK b6252655fe502ed1bbb97ab9597d9bbcf8c4484c

This PR adds capital H to the parser as a valid Hardened indicator for the descriptor. Capital H usage as an indicator is per BIP-380 spec. This makes the ParseKeyPathNum function adhere (more) to the BIP spec.   

- Code review ✅
- Build and tested ✅
- Checked that the test fails if the addition to the parser is removed. ✅

rkrux commented on src/test/descriptor_tests.cpp:
```diff
@@ -554,6 +560,7 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
     Check("combo([01234567]xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc)", "combo([01234567]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL)", "combo([01234567]xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvR ...
     Check("pk(xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0)", "pk(xpub68NZiKmJWnxxS6aaHmn81bvJeTESw724CRDs6HbuccFQN9Ku14VQrADWgqbhhTHBaohPX4CjNLf9fq9MYo6oDaPPLPxSb7gwQN3ih19Zm4Y/0)", "pk(xpub68NZiKmJWnxxS6aaHmn81bvJeTESw724CRDs6HbuccFQN9Ku14VQrADWgqbhhTHBaohPX4CjNLf9fq9MYo6oDaPPLPxSb7gwQN3ih19Zm4Y/0)", DEFAULT, {{"210379e45b3cf7 ...
     Check("pkh(xprv9s21ZrQH143K31xYSDQpPDxsXRTUcvj2iNHm5NUtrGiGG5e2DtALGdso3pGz6ssrdK4PFmM8NSpSBHNqPqm55Qn3LqFtT2emdEXVYsCzC2U/2147483647'/0)", "pkh(xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB/2147483647'/0)", "pkh([bd16bee5/2147483647h]xpub69H7F5dQzmVd3vPuLKtcXJziMEQByuDidnX3YdwgtNsecY5HRGtAAQC5mXTt4dsv9RzyjgDjAQs9VGVV6ydYCHnprc ...
+    Check("pkh([deadbeef/1H/1'/1h]xprv9s21ZrQH143K31xYSDQpPDxsXRTUcvj2iNHm5NUtrGiGG5e2DtALGdso3pGz6ssrdK4PFmM8NSpSBHNqPqm55Qn3LqFtT2emdEXVYsCzC2U/2147483647H/0h/0'/0)", "pkh([deadbeef/1H/1'/1h]xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB/2147483647H/0h/0'/0)", "pkh([deadbeef/1h/1h/1h/2147483647h/0h/0h]xpub6D9zrB5MDjRMPNekz6itAEnsE ...
```
+1 on passing the max int value in the test.

rkrux reviewed (commented):
Concept ACK b6252655fe502ed1bbb97ab9597d9bbcf8c4484c

I feel the following site should also be updated for uniformity in this PR. 

Nit: This condition can also be extracted out in a `IsHardenedIndicator(char input)` utility that can be reused to avoid any misses in the future.

https://github.com/bitcoin/bitcoin/blob/c5849663baa925a5291731e4e4013a08c811c646/src/test/descriptor_tests.cpp#L495

Sjors commented:
Concept ACK

Can you modify one of the functional tests to cover the use of `H` in `importdescriptors` (e.g. `wallet_signer.py`) and `getdescriptorinfo`?

rkrux reviewed (changes requested):
In addition to this comment https://github.com/bitcoin/bitcoin/pull/32788#pullrequestreview-2949654103,

this site for hardened range derivation should also be updated inside `ParsePubkeyInner` that is called via descriptor `Parse`, I think a common utility function here could prove to be robust. Maybe a functional test change as suggested in this https://github.com/bitcoin/bitcoin/pull/32788#issuecomment-2996082521 could have caught this discrepancy.
https://github.com/bitcoin/bitcoin/blob/c5849663baa925a5291731e4e4013a08c811c646/src/script/descriptor.cpp#L1567

Also, updated comments in the following places would be good to have:
```bash
➜  bitcoin git: git grep -n "' or h" src/script/descriptor.cpp
src/script/descriptor.cpp:374:    // Whether ' or h is used in harded derivation
src/script/descriptor.cpp:1648: * @param[in] split BIP32 path string, using either ' or h for hardened derivation
```

rkrux commented:
> Why is another way to specify hardened derivation needed? Is anyone using it? Seems better to remove H from BIP.

I would prefer the removal from BIP if no other software is using it but I'm not sure if there is a way to comprehensively verify if this has not been used in the past. And in the lack of such a verification, the allowance in core wallet seems required to me.

achow101 commented:
> How those "softwares" worked up until this point?

We don't know, which is kind of the point of this PR.

The text of the BIP has always stated that `H` is valid hardened indicator, so anyone implementing from just the description of the text should have included `H`.

The document the BIP is based on (doc/descriptors.md) did not state that `H` is a valid hardened indicator.

However, there was also a test vector that had `H` as an invalid expression. Anyone who also implemented this test vector would have noticed that the text of the BIP did not match what this test vector stated was invalid. But this test vector was also added to the BIP about 2 years after the BIP was originally written.

Since this discrepancy was only flagged a few months ago, it seems like there are multiple possibilities of what could have happened:

1. Many implementations of descriptors was done prior to the test vectors, and probably prior to the BIP being written, as we know descriptors was specified long before the BIP. Such implementations may or may not have implemented it correctly.
2. Some implementations of descriptors were done after the test vectors were added, but did not implement the test vectors, or implemented based off of existing implementations which may or may not have allowed `H`.
3. Few implementations were done specifically from the text and tests alone, as no one flagged this discrepancy for multiple years.

The 2 primary implementations of descriptors are probably Bitcoin Core and rust-miniscript. Looking at rust-miniscript, it seems that they also did not allow `H`. However, even if people tested against Bitcoin Core and rust-bitcoin, it's possible that they could be allowing `H` as they may not be producing descriptors with `H`. For example, it looks like libwally allows `H` as a hardened indicator, so any implementation that uses libwally for handling things like extended keys and derivation paths are probably allowing `H`.

Maybe I'll ask on the mailing list again.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32788/commits/b6252655fe502ed1bbb97ab9597d9bbcf8c4484c

pythcoiner commented:
> Is anyone using it?

support for `H` have been introduced in embit in [this commit](https://github.com/diybitcoinhardware/embit/pull/47/commits/1696d6815cf6c2ddb7f56dfbf3bc50162e463647) but afair I add `H` because it was in the BIP not because it was needed.

achow101 commented:
https://github.com/bitcoin/bips/pull/1888 was merged so this is no longer relevant.
